src/components/fileviewer.js

```diff
@@ -165,3 +170,32 @@ const FileViewerMeta = ({ revision, path, appError }) => {
     </div>
   );
 };
+
+const CoverageMeta = ({ coverage }) => {
+  // if there is no coverage information, then do not display the percentage covered
+  let percentageCovered = undefined;
+  
+  if (coverage) {
+    let totalCovered = _.union(_.flatten(coverage.data.map((d) => d.source.file.covered)));
+    let uncovered = _.union(_.flatten(coverage.data.map((d) => d.source.file.uncovered)));
+    let totalUncovered = _.difference(uncovered, totalCovered);
+    // if there is coverage information and no lines are uncovered, percentage covered is 100%
+    if (totalCovered.length > 0 && totalUncovered.length === 0) {
+      this.percentageCovered = 1;
+    }
+    // calculate the percentage covered if there is coverage information
+    if (!(totalCovered === 0 && uncovered === 0)) {
+      this.percentageCovered = totalCovered.length / totalUncovered.length;
+    }
```

Comment:
It is hard to confirm the correctness of this code.

    if (totalCovered.length > 0 && totalUncovered.length === 0) {
      this.percentageCovered = 1;
    }
    // calculate the percentage covered if there is coverage information
    if (!(totalCovered === 0 && uncovered === 0)) {
      this.percentageCovered = totalCovered.length / totalUncovered.length;
    }

In general, if you are conditionally setting one variable (`this.percentageCovered`), write your code so that it is easy to confirm it will be set, no matter what path through the code is taken. For example, it is hard for me to verify that `this.percentageCovered` is always set; is there a situation that misses both `if` blocks?

The  two `if` blocks can be merged into one `if/else` block.

    // calculate the percentage covered if there is coverage information
    if (!(totalCovered === 0 && uncovered === 0)) {
      this.percentageCovered = totalCovered.length / totalUncovered.length;
    } else {
      this.percentageCovered = 1;
    }

Please use Boolean algebra to convert a negative statement into a positive one; pushing the negation deeper into the expression:  `!(totalCovered === 0 && uncovered === 0)` can be restated as `(totalCovered !=0 or uncovered != 0)`.

Finally, the denominator is wrong (which i also go wrong): It should be the total number of lines.

    this.percentageCovered = totalCovered.length / (totalCovered.length + totalUncovered.length)

Which makes me wonder if `totalUncovered` is ever needed; instead calculate `totalLines` early.

    let totalLines = _.union(uncovered, totalCovered).length

Reply:
The way it is currently is that percentageCovered isn't always set as in the case when totalCovered === 0 && uncovered === 0.  I'm not sure if we wanted the percentage to be calculated and shown when we didn't have any coverage information.  The case where totalCovered.length > 0 && totalUncovered.length === 0 is when we only have covered lines and no uncovered lines, then we need to account for division by 0.  I'm not sure if I'm interpreting this right and if these are realistic cases however.

Reply:
OK, if there is a situation that `percentageCovered` should not be set, then the same rule applies; do your assignment in a single `if/else` block, and be explicit about not setting it.

```javascript
    if (!(totalCovered === 0 && uncovered === 0)) {
      this.percentageCovered = totalCovered.length / ( totalCovered.length + totalUncovered.length);
    } else {
      //this.percentageCovered is left undefined
    }
```

> The case where totalCovered.length > 0 && totalUncovered.length === 0 is when we only have covered lines and no uncovered lines, then we need to account for division by 0.

On this point, there is no division by zero: The denominator is `totalCovered.length + totalUncovered.length`.